include/signal.h

```diff
@@ -0,0 +1,50 @@
+#ifndef _SYS_SIGNAL_H_
+#define _SYS_SIGNAL_H_
+
+#include <common.h>
+#include <bitstring.h>
+#include <queue.h>
+#include <mutex.h>
+#include <signum.h>
+
+typedef struct proc proc_t;
+
+typedef bitstr_t sigset_t[bitstr_size(SIG_LAST)];
+
+typedef void (*sighandler_t)(int);
+
+#define SIG_DFL (sighandler_t *)0x00
+#define SIG_IGN (sighandler_t *)0x01
+#define SIG_TERM (sighandler_t *)0x02
```

Comment:
Eh, kinda. Let's look at `signal.h` in Linux. This description may be slightly inaccurate for some distros, though what I explain is true for most of them. The basic `/usr/include/signal.h` is provided by `glibc`, and it usually doesn't do much except for including `/usr/include/sys/signal.h` (sometimes it's the other way round, but anyway these files end up being equivalent). But `sys/signal.h`, and any file in `/usr/include/sys` for that matter, does not come directly from kernel sources. It is also provided by `glibc`, where it was written from scratch. But the files in `sys` (mostly) provide just the function interface, not the values of constants. These are pulled in by including files from `/usr/include/bits` directory. These files are (usually) forbidden to be included directly. These expose the *structure layout* and *constant values* for use with system interfaces. Great! So are these the files provided by kernel? No. These are also maintained by `glibc` developers.

So what *does* the kernel provide? `/usr/include/linux`. It does contain a `signal.h`, which pulls in stuff from `/usr/include/asm{,-generic}/signal*.h`. These contain mostly the same definitions as `bits/signal.h`, but these are maintained by kernel developers. Applications don't typically use interfaces from `/usr/include/linux` unless they need linux-specific features.

I recall that `uClibc` standard C library is capable of generating the `/usr/include/bits` files from `/usr/include/linux` (if we were to provide these), which seems convenient. Obviously, `newlib` is not that cool (it really doesn't care about a potentially present operating system).